File: test/lit/zig/kitchen_sink.zig
// RUN: %cot emit-cir %s | %FileCheck %s

// Kitchen sink: all Zig features COT supports in one real-world example.
// This must be valid Zig — compilable by zig build-exe.

const Vec2 = struct {
    x: i32,
    y: i32,
};

const MathError = error{ Overflow, DivByZero };

pub fn vec2Add(a: Vec2, b: Vec2) Vec2 {
    return Vec2{ .x = a.x + b.x, .y = a.y + b.y };
}

pub fn dot(a: Vec2, b: Vec2) i32 {
    return a.x * b.x + a.y * b.y;
}

pub fn magnitude_squared(v: Vec2) i32 {
    return dot(v, v);
}

pub fn safeDivide(a: i32, b: i32) MathError!i32 {
    if (b == 0) return error.DivByZero;
    return @divTrunc(a, b);
}

pub fn clamp(val: i32, lo: i32, hi: i32) i32 {
    if (val < lo) return lo;
    if (val > hi) return hi;
    return val;
}

pub fn sumArray(arr: [4]i32) i32 {
    var total: i32 = 0;
    var i: i32 = 0;
    while (i < 4) {
        total = total + arr[@intCast(i)];
        i += 1;
    }
    return total;
}

pub fn firstOrNull(opt: ?i32) i32 {
    if (opt) |val| {
        return val;
    }
    return 0;
}

pub fn greetLen(s: []const u8) i64 {
    return s.len;
}

// CHECK-LABEL: func.func @vec2Add
// CHECK: cir.field_val
// CHECK: cir.add
// CHECK: cir.struct_init

// CHECK-LABEL: func.func @dot
// CHECK: cir.mul
// CHECK: cir.add

// CHECK-LABEL: func.func @magnitude_squared
// CHECK: call @dot

// CHECK-LABEL: func.func @safeDivide
// CHECK: cir.cmp eq
// CHECK: cir.wrap_error
// CHECK: cir.div
// CHECK: cir.wrap_result

// CHECK-LABEL: func.func @clamp
// CHECK: cir.cmp slt
// CHECK: cir.cmp sgt

// CHECK-LABEL: func.func @sumArray
// CHECK: cir.alloca
// CHECK: cir.cmp slt
// CHECK: cir.condbr
// CHECK: cir.add

// CHECK-LABEL: func.func @firstOrNull
// CHECK: cir.is_non_null
// CHECK: cir.optional_payload

// CHECK-LABEL: func.func @greetLen
// CHECK: cir.slice_len

File: test/lit/ts/kitchen_sink.ts
// RUN: %cot emit-cir %s | %FileCheck %s

// Kitchen sink: all TypeScript features COT supports in one real-world example.
// This must be valid TypeScript — compilable by tsc.

interface Vec2 {
    x: number;
    y: number;
}

function vec2Add(a: Vec2, b: Vec2): Vec2 {
    return { x: a.x + b.x, y: a.y + b.y };
}

function dot(a: Vec2, b: Vec2): number {
    return a.x * b.x + a.y * b.y;
}

function magnitudeSquared(v: Vec2): number {
    return dot(v, v);
}

function clamp(val: number, lo: number, hi: number): number {
    if (val < lo) return lo;
    if (val > hi) return hi;
    return val;
}

function abs(x: number): number {
    return x > 0 ? x : 0 - x;
}

function sumTo(n: number): number {
    let total: number = 0;
    for (let i: number = 0; i < n; i = i + 1) {
        total = total + i;
    }
    return total;
}

function factorial(n: number): number {
    let result: number = 1;
    let i: number = 2;
    while (i <= n) {
        result = result * i;
        i += 1;
    }
    return result;
}

function isEven(n: number): boolean {
    return n % 2 == 0;
}

function greetLen(s: string): number {
    return s.len;
}

// CHECK-LABEL: func.func @vec2Add
// CHECK: cir.field_val
// CHECK: cir.add
// CHECK: cir.struct_init

// CHECK-LABEL: func.func @dot
// CHECK: cir.mul
// CHECK: cir.add

// CHECK-LABEL: func.func @magnitudeSquared
// CHECK: call @dot

// CHECK-LABEL: func.func @clamp
// CHECK: cir.cmp slt
// CHECK: cir.cmp sgt

// CHECK-LABEL: func.func @abs
// CHECK: cir.cmp sgt
// CHECK: cir.select

// CHECK-LABEL: func.func @sumTo
// CHECK: cir.cmp slt
// CHECK: cir.condbr
// CHECK: cir.add

// CHECK-LABEL: func.func @factorial
// CHECK: cir.cmp sle
// CHECK: cir.mul

// CHECK-LABEL: func.func @isEven
// CHECK: cir.rem
// CHECK: cir.cmp eq

// CHECK-LABEL: func.func @greetLen
// CHECK: cir.slice_len
